backend/python/tapa/floorplan_dse.py

```diff
@@ -0,0 +1,201 @@
+      for regexp in v:
+        cmd += [to_tapac_option(k), regexp]
+    # no conversion needed
+    else:
+      cmd += [to_tapac_option(k), str(v)]
+
+  # the last option is the input source file
+  cmd += [args_dict['input_file']]
+
+  return cmd
+
+
+def get_cmd_to_csynth_tasks(
+    args_dict: Dict[str, Any],
+    work_dir: str,
+    output_dir: str,
+) -> List[str]:
+  """Before the actual DSE, run one pass of hls synthesis and enable_synth_util"""
+  basic_cmd = get_basic_tapa_command(args_dict, output_dir)
+  basic_cmd += [
+    '--work-dir', work_dir,
+    '--run-tapacc',
+    '--run-hls',
+    '--generate-task-rtl',
+  ]
+  return basic_cmd
+
+
+def get_cmd_to_explore_floorplan(
+    args_dict: Dict[str, Any],
+    work_dir: str,
+    output_dir: str,
+    max_slr_limit: int,
+    min_area_limit: float,
+) -> List[str]:
+  """Command to get one design point. Reuse HLS synthesis and enable_synth_util"""
+  basic_cmd = get_basic_tapa_command(args_dict, output_dir)
+  basic_cmd += [
+    '--work-dir', work_dir,
+    '--generate-task-rtl',
+    '--run-floorplanning',
+    '--generate-top-rtl',
+    '--pack-xo',
+    '--max-slr-width-limit', str(max_slr_limit),
+    '--min-area-limit', str(min_area_limit),
+  ]
+  return basic_cmd
+
+
+def run_floorplan_dse(args):
+  """Self-invoke tapac to generate multiple floorplan outputs"""
+  # remove the file opened in argparse
+  args.floorplan_output.close()
+  os.remove(args.floorplan_output.name)
+
+  args_dict = vars(args)
+  args_dict = {k:v for k, v in args_dict.items() if v is not None}
+
+  top_work_dir = args.work_dir
+  try:
+    os.makedirs(top_work_dir, exist_ok=True)
+  except:
+    _logger.error('Fail to make directory at %s. Check the --work-dir option.', top_work_dir)
+    exit(1)
+
+  _logger.info(f'')
+  _logger.info(f'-----------------------------------------------------------------------------')
+  _logger.info(f'DSE preparation: synthesize each task into RTL')
+  _logger.info(f'-----------------------------------------------------------------------------')
+  _logger.info(f'')
+
+  hls_dir = f'{os.getcwd()}/{top_work_dir}/.csynth/'
+  hls_cmd = get_cmd_to_csynth_tasks(args_dict, hls_dir, tempfile.gettempdir())
+  try:
+    subprocess.run(hls_cmd,
+                    stdout=subprocess.PIPE,
+                    check=True,
+                    universal_newlines=True)
+  except subprocess.CalledProcessError as e:
+    _logger.error('Fail to run command: %s', ' '.join(hls_cmd) )
+    exit(1)
+
+  # start exploring the pareto optimal curve
+  min_slr_limit = args.min_slr_width_limit
+  max_slr_limit = args.max_slr_width_limit
+  min_area_limit = args.min_area_limit
+  id = 1
+
+  while min_slr_limit < max_slr_limit:
+    _logger.info(f'')
+    _logger.info(f'-----------------------------------------------------------------------------')
+    _logger.info(f'Start generating design point {id} with --max-slr-width-limit {max_slr_limit}')
```

Comment:
What's the rationale for these lines?

Reply:
Since we are self-calling tapac, it looks better to have some clear separation in the logs to differentiate the parts that belong to different tapac calls. Other suggestions?

Reply:
How about suppressing the logging level for subprocesses? It can help reduce log spamming and help people focus on important messages. We can set the logging level of the log file to always be DEBUG so that if anything goes wrong, we always have a detailed trace.

Reply:
Ideally we should make the steps be functions with well-defined inputs and outputs so that subprocesses won't be necessary, and the logging messages would be more coherent. However, migration can take a lot of time. I'm fine with subprocesses with some repetitive logging.